File: algorithm/BaekJoon/BaekJoon_1916.java
package BaekJoon;

import java.io.*;
import java.util.*;

public class BaekJoon_1916 {
    static int INF = Integer.MAX_VALUE;
    static List<node1>[] list;
    static boolean check[];
    static int[] distance;
    
    @SuppressWarnings("unchecked")

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());

        int m = Integer.parseInt(st.nextToken());
        int start;
        int last;

        check = new boolean[n+1];
        distance = new int[n+1];
        list = new ArrayList[n+1];

        Arrays.fill(distance, INF);
    
        for(int i = 1; i <= n; i++){
            list[i] = new ArrayList<>();
        }

        for(int i = 0; i < m; i++){
            st = new StringTokenizer(br.readLine());

            while(st.hasMoreTokens()){
                int index = Integer.parseInt(st.nextToken());
                int finish = Integer.parseInt(st.nextToken());
                int weight = Integer.parseInt(st.nextToken());

                list[index].add(new node1(finish, weight));
            }
        }
        st = new StringTokenizer(br.readLine());

        start = Integer.parseInt(st.nextToken());
        last = Integer.parseInt(st.nextToken());

        dijkstra(start, last);

        System.out.println(distance[last]);

    }

    private static void dijkstra(int s, int l){
        PriorityQueue<node1> q = new PriorityQueue<>();

        q.add(new node1(s, 0));
        distance[s] = 0;

        while(!q.isEmpty()){
            node1 node = q.poll();
            int now = node.index;
            

            if(check[now] == true) continue;
            check[now] = true;

            for(node1 n : list[now]){
                if(distance[n.index] > distance[now] + n.weight){
                    distance[n.index] = distance[now] + n.weight;
                    q.add(new node1(n.index, distance[n.index]));
                }
            }
        }
    }
    
}

class node1 implements Comparable<node1>{
    int index;
    int weight;

    node1(int i, int w){
        this.index = i;
        this.weight = w;
    }

    @Override
    public int compareTo(node1 o) {
        return weight - o.weight;
    }
}

File: algorithm/BaekJoon/Kotlin/BaekJoon_1916.kt
import BaekJoon_1916.Companion.check
import BaekJoon_1916.Companion.distance
import BaekJoon_1916.Companion.list
import java.io.*
import java.util.*
import kotlin.collections.ArrayList


class BaekJoon_1916 {
    companion object {
        lateinit var distance: Array<Int>
        lateinit var check: Array<Boolean>
        lateinit var list : Array<ArrayList<Node>>
    }
}

class Node : Comparable<Node>{
    var index : Int = 0
    var weight : Int = 0

    constructor(i : Int, w : Int){
        this.index = i
        this.weight = w
    }


    override fun compareTo(other: Node): Int {
        return this.weight - other.weight
    }

}

fun main(){
    var br = BufferedReader(InputStreamReader(System.`in`))
    var n = br.readLine().toInt()
    var m = br.readLine().toInt()

    list = Array(n+1){ArrayList<Node>()}

    for(i in 0..n){
        list[i] = ArrayList()
    }

    check = Array(n+1){false}
    distance = Array(n+1){Integer.MAX_VALUE}

    for(i in 0 until m){
        var(index,finish,weight) = br.readLine().split(' ').map { it.toInt() }
        list[index].add(Node(finish, weight))
    }

    var(start, finish) = br.readLine().split(' ').map { it.toInt() }

    dijk(start)

    println(distance[finish])

}

fun dijk(start : Int){
    var pq = PriorityQueue<Node>()

    pq.add(Node(start, 0))
    distance[start] = 0

    while(!pq.isEmpty()){
        var n : Node = pq.poll()
        var now : Int = n.index

        if(check[now] == true) continue
        check[now] = true

        for(n in list[now]){
            if(distance[n.index] > distance[now] + n.weight){
                distance[n.index] = distance[now] + n.weight
                pq.add(Node(n.index, distance[n.index]))
            }
        }
    }
}